docs/building-modules/upgrade.md

```diff
@@ -0,0 +1,42 @@
+<!--
+order: 13
+-->
+
+# Upgrade
+
+The Upgrade module allows your modules to smoothly transition to new versions with breaking changes. This document outlines how to build modules to take advantage of this functionality. 
+
+# Consensus Version
+
+In order to successfully upgrade your existing modules, your `AppModule`s must implement the function `ConsensusVersion() uint64`. The `uint64` returned will serve as the consensus version and should be hard coded by the module developer. This number will serve as a state-breaking version of each app module. The initial version *MUST* be set to 1.
+
+# Registering Migrations
+
+To register the functionality that takes place during a module upgrade, we must register which migrations we want to take place. This takes place in the `Configurator` via the `RegisterMigration` method. The `AppModule`s have a reference to the configurator in the `RegisterServices` method. We can register a number of migrations here, however, if more than one migration script is registered, it is important they are listed in increasing order. Additionally, we must ensure there are enough migrations that will lead to the desired consensus version. For example, if we wanted to migrate to version 3 of a module, we would need to register a separate migration for both version 1 and 2 as shown below.
+
+```golang
+func (am AppModule) RegisterServices(cfg module.Configurator) {
+    // --snip--
+    cfg.RegisterMigration(types.ModuleName, 1, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 1 to 2.
+    })
+     cfg.RegisterMigration(types.ModuleName, 2, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 2 to 3.
+    })
+}
+```
+
+Since these migrations are functions that need access to a Keeper's store, we use a wrapper around the keepers called `Migrator`. An example of this can be found (here)[https://github.com/cosmos/cosmos-sdk/blob/master/x/bank/keeper/migrations.go]. In addition to the `Migrator` wrapper, we also define our migration scripts. More on that below.
```

Comment:
```suggestion
Since these migrations are functions that need access to a Keeper's store, we use a wrapper around the keepers called `Migrator`. An example of this can be found here:

+++ https://github.com/cosmos/cosmos-sdk/blob/6ac8898fec9bd7ea2c1e5c79e0ed0c3f827beb55/x/bank/keeper/migrations.go#L8-L21

In addition to the `Migrator` wrapper, we also define our migration scripts. More on that below.
```

vuepress has this `+++` markdown syntax which allows pasting a sniippet of code in the docs. I think it could be useful here for module developers, wdyt?

Also, we try to put permalinks in docs, so that links don't break.